Send deflated DICOM file via C-STORE

amoerie opened the pull request:
Fixes #1281
Fixes #1282

#### Checklist
- [x] The pull request branch is in sync with latest commit on the *fo-dicom/development* branch
- [x] I have updated API documentation
- [x] I have included unit tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file

#### Changes proposed in this pull request:
- When the accepted presentation context specifies that the transfer syntax is deflated, automatically deflate the byte content that we send
- While the sending doesn't go wrong on our side, we are otherwise sending byte contents to the receiver inflated, while telling them it is deflated. This will always cause an error if the receiving party then actually tries to inflate.
- Example unit test is included that sends a deflated DICOM file from DicomClient -> DicomServer

amoerie committed:
Add bug that demonstrates the deflation issue

amoerie committed:
Enforce max length in DicomStatus

Fixes #1282

amoerie committed:
Automatically deflate DICOM dataset

If the accepted transfer syntax requires deflation, automatically deflate the DICOM data set when sending it over the wire

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into GH-1282

amoerie committed:
Include deflated DICOM file sending fix in changelog

amoerie committed:
Merge branch 'development' into GH-1282

amoerie committed:
Increase association request timeout in tests

This should help avoid flakiness when we are over our CI budget and are running tests on slow machines

amoerie commented:
Bonus question: should we also propose "Explicit VR Little Endian" automatically when sending a "Deflated Explicit VR Little Endian"? Now we only propose "Deflated VR Little Endian" or "Implicit VR Little Endian", but if the receiving party does not supported deflated DICOM files, it HAS to choose implicit, in which case the private tag VRs will be lost.

I would suggest to also always include transfer syntax "Explicit VR Little Endian" here. This is what we currently send:

```
Presentation Contexts:  2
  Presentation Context:  1 [Proposed]
       Abstract Syntax:  Unknown [2.25.152752519834663263826301420581341174218]
       Transfer Syntax:  Deflated Explicit VR Little Endian
  Presentation Context:  3 [Proposed]
       Abstract Syntax:  Unknown [2.25.152752519834663263826301420581341174218]
       Transfer Syntax:  Implicit VR Little Endian: Default Transfer Syntax for DICOM
```

gofal commented:
I also think, the sentence in DICOM standard makes it clear. fo-dicom should also offer Explicit Little Endian in case of Deflated Explicit Little Endian.

amoerie committed:
Add Explicit VR Little Endian transfer syntax

When sending a deflated DICOM file via C-STORE, always include Explicit VR Little Endian in the proposed transfer syntaxes too, as required by the DICOM standard

amoerie committed:
Add JetBrains Rider/Resharper team shared settings file

This contains IDE related settings (for those who use Rider/Resharper) that are shared across the team

For now, the file is rather empty, but it can be useful in the future to consolidate certain settings that are beyond the default editor config possibilities.

amoerie committed:
Add specific tests for DicomPresentationContextCollection

amoerie commented:
I implemented the automatic suggestion of explicit VR Little Endian, added tests specifically for the DicomPresentationContextCollection.AddFromRequest method. 

But now the TempFileBuffer test is failing again. 😫

mrbean-bremen commented on FO-DICOM.Full.sln.DotSettings:
```diff
@@ -0,0 +1,3 @@
+﻿<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
+	<s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/Abbreviations/=GH/@EntryIndexedValue">GH</s:String>
```
This only adds `GH` to the dictionary, or does it do something else?

amoerie replied:
Yes it just tells Resharper/Rider to not suggest Gh instead of GH when naming a class. It's absolutely trivial, but I like to put little IDE related things in source code for everyone.

mrbean-bremen commented on Tests/FO-DICOM.Tests/Network/AsyncDicomCFindProviderTests.cs:
```diff
@@ -38,6 +38,7 @@ public async Task OnCFindRequestAsync_ImmediateSuccess_ShouldRespond()
             {
                 var client = DicomClientFactory.Create("127.0.0.1", port, false, "SCU", "ANY-SCP");
                 client.Logger = _logger.IncludePrefix(typeof(DicomClient).Name);
+                client.ClientOptions.AssociationRequestTimeoutInMs = (int) TimeSpan.FromMinutes(5).TotalMilliseconds;
```
Have these been flaky tests, or is it somehow related to the changes?

amoerie replied:
Yes, from what I have seen, it looks like GitHub actions puts us on slower build machines if we trigger too many builds in a day.

On .NET framework, the whole async Task story is less optimized than in .NET core, and I believe we suffer from thread pool starvation issues when running the tests. (I.e. a Task has produced a result but there is no free thread on the thread pool to run the continuation)

This means that the association request acceptance continuation takes longer than 5 seconds (the default timeout) to execute (even though the server accepts immediately) and then the test fails due to timeouts. 

I think having much bigger timeouts is no big issue here, especially because this thread starvation issue will only get worse when we add more tests. If we see this happening more and more, we might have to investigate adding more tests into a single xUnit collection to reduce the total parallelization.

mrbean-bremen replied:
Thanks, that makes sense! I also think that .NET framework will not be used much for the server part - it can easily be made a separate .Net Core application, even if the frontend still uses framework, So hopefully this will not be a big deal in the near future.

amoerie replied:
> Thanks, that makes sense! I also think that .NET framework will not be used much for the server part - it can easily be made a separate .Net Core application, even if the frontend still uses framework, So hopefully this will not be a big deal in the near future.

I think this issue might not even exist in real life circumstances. The tests are spinning up hundreds of DICOM servers, sending thousands of requests in some cases, dropping connections, transcoding images, etc. Running the fo-dicom tests is very intensive. I imagine real life scenarios are very different (only a few servers, for example)

mrbean-bremen commented on Tests/FO-DICOM.Tests/Network/DicomPresentationContextCollectionTests.cs:
```diff
@@ -0,0 +1,116 @@
+        {
+            // Arrange
+            var presentationContexts = new DicomPresentationContextCollection();
+            var file = "Test Data/CR-ModalitySequenceLUT.dcm";
```
Maybe add a comment with the transfer syntax this has (I had to check that it is not already encoded as Implicit Little Endian).

amoerie replied:
Good idea, will do.

amoerie replied:
Done

mrbean-bremen reviewed (approved):
Looks good to me! (with the usual disclaimer that I'm not familiar enough with fo-dicom)

mrbean-bremen commented:
Again the failing test for temporary files...

amoerie committed:
Document the original transfer syntax in tests

amoerie committed:
Merge development into GH-1282
